File: scripts/truncate_embeddings.py
"""Truncate an embeddings.bin to a smaller dimension using Matryoshka-style
prefix truncation + renormalization, instead of re-calling the embeddings API.

Usage: python3 truncate_embeddings.py <data_dir> <meta_filename> [dst_dim]
e.g.   python3 truncate_embeddings.py ../data/plan_policies chunks-meta.json 256
       python3 truncate_embeddings.py ../data/ecm_guide obligations-meta.json 256
"""
import json
import math
import os
import struct
import sys

SRC_DIM = 768


def normalize(vec):
    norm = math.sqrt(sum(x * x for x in vec))
    return [x / norm for x in vec] if norm > 0 else vec


def main():
    data_dir = sys.argv[1]
    meta_filename = sys.argv[2]
    dst_dim = int(sys.argv[3]) if len(sys.argv) > 3 else 256

    meta_file = os.path.join(data_dir, meta_filename)
    out_file = os.path.join(data_dir, "embeddings.bin")
    backup_path = os.path.join(data_dir, f"embeddings-{SRC_DIM}d.bak.bin")
    in_file = backup_path if os.path.exists(backup_path) else out_file

    with open(meta_file) as f:
        meta = json.load(f)

    with open(in_file, "rb") as f:
        src = f.read()

    count = len(src) // (SRC_DIM * 4)
    if len(src) % (SRC_DIM * 4) != 0:
        raise ValueError(f"embeddings.bin size doesn't divide evenly by SRC_DIM={SRC_DIM}")
    if count != len(meta):
        raise ValueError(f"vector count ({count}) != meta entries ({len(meta)})")

    if os.path.abspath(in_file) != os.path.abspath(backup_path):
        with open(backup_path, "wb") as f:
            f.write(src)
        print(f"Backed up original to {backup_path} (not tracked by git, delete when done)")

    with open(out_file, "wb") as out:
        for i in range(count):
            vec = struct.unpack_from(f"<{SRC_DIM}f", src, i * SRC_DIM * 4)[:dst_dim]
            vec = normalize(list(vec))
            out.write(struct.pack(f"<{dst_dim}f", *vec))

    print(
        f"Wrote {out_file}: {count} vectors, {SRC_DIM}d -> {dst_dim}d "
        f"({count * dst_dim * 4 / 1024 / 1024:.1f}MB, was {len(src) / 1024 / 1024:.1f}MB)"
    )


if __name__ == "__main__":
    main()

File: scripts/truncate-embeddings.ts
import fs from "fs";
import path from "path";

const BACKUP_FILE = path.join(process.cwd(), "data", "embeddings-768d.bak.bin");
const IN_FILE = fs.existsSync(BACKUP_FILE)
  ? BACKUP_FILE
  : path.join(process.cwd(), "data", "embeddings.bin");
const OUT_FILE = path.join(process.cwd(), "data", "embeddings.bin");
const META_FILE = path.join(process.cwd(), "data", "chunks-meta.json");

const SRC_DIM = 768;
const DST_DIM = process.argv[2] ? parseInt(process.argv[2], 10) : 256;

function normalize(v: Float32Array): Float32Array {
  let norm = 0;
  for (let i = 0; i < v.length; i++) norm += v[i] * v[i];
  norm = Math.sqrt(norm);
  if (norm > 0) for (let i = 0; i < v.length; i++) v[i] /= norm;
  return v;
}

function main() {
  const meta = JSON.parse(fs.readFileSync(META_FILE, "utf-8"));
  const srcBuffer = fs.readFileSync(IN_FILE);
  const count = srcBuffer.length / (SRC_DIM * 4);

  if (!Number.isInteger(count)) {
    throw new Error(`embeddings.bin size doesn't divide evenly by SRC_DIM=${SRC_DIM}`);
  }
  if (count !== meta.length) {
    throw new Error(`vector count (${count}) != meta entries (${meta.length})`);
  }

  const backupPath = path.join(process.cwd(), "data", `embeddings-${SRC_DIM}d.bak.bin`);
  if (path.resolve(IN_FILE) !== path.resolve(backupPath)) {
    fs.copyFileSync(IN_FILE, backupPath);
    console.log(`Backed up original to ${backupPath} (not tracked by git, delete when done)`);
  }

  const dstBuffer = Buffer.alloc(count * DST_DIM * 4);
  for (let i = 0; i < count; i++) {
    const vec = new Float32Array(DST_DIM);
    for (let j = 0; j < DST_DIM; j++) {
      vec[j] = srcBuffer.readFloatLE((i * SRC_DIM + j) * 4);
    }
    normalize(vec);
    for (let j = 0; j < DST_DIM; j++) {
      dstBuffer.writeFloatLE(vec[j], (i * DST_DIM + j) * 4);
    }
  }

  fs.writeFileSync(OUT_FILE, dstBuffer);
  console.log(
    `Wrote ${OUT_FILE}: ${count} vectors, ${SRC_DIM}d -> ${DST_DIM}d ` +
      `(${(dstBuffer.length / 1024 / 1024).toFixed(1)}MB, was ${(srcBuffer.length / 1024 / 1024).toFixed(1)}MB)`
  );
}

main();
